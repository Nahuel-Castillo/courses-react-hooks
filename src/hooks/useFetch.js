import { useState, useEffect, useRef } from 'react';


export const useFetch = ( url ) => {
    
    const [state, setState] = useState({ data: null, loading: true, error: null });

    const isMounted = useRef(true);
    
    useEffect(() => {
        return () => {
            isMounted.current = false;    
        }
    }, []);

    useEffect( () => {

        setState({
            data: null, 
            loading: true,
            error: null
        });

        fetch( url )
            .then( resp => resp.json() )
            .then( data => {

                if ( isMounted ) {

                    setState({
                        loading: false,
                        error: null,
                        data
                    });
                }
            })
            .catch( () => {

                setState({
                    loading: false,
                    data: null,
                    error: 'Can not load data'
                });
            });

    }, [ url ]);

    return state;
}
