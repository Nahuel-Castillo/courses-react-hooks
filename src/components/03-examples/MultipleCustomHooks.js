import React, { useState } from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

import '../02-useEffect/effects.css';

export const MultipleCustomHooks = () => {

    const { counter, increment, decrement} = useCounter(1);

    const { data, loading } = useFetch(`https://www.breakingbadapi.com/api/quotes/${ counter < 1 ? 1 : counter }`);

    const { author, quote } = !loading && data[0];

    return (
        <div>
            <h2> BrekingBad Quotes </h2>
            <hr/>

            { 
                loading
                ?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>
                    )
                : 
                    (
                        <>
                            <blockquote className="blockquote text-right">
                                <p className="mb-3"> { quote } </p>
                                <footer className="blockquote-footer"> { author }  </footer>
                            </blockquote>
                            <hr/>

                            <button 
                                onClick={ decrement } 
                                className="btn btn-primary m-2" 
                                disabled={ counter <= 1 }
                            >
                                Previous quote
                            </button> 

                            <button onClick={ increment } className="btn btn-primary m-2">
                                Next quote
                            </button> 
                        </>
                    ) 
            }
        </div>
    )
}
