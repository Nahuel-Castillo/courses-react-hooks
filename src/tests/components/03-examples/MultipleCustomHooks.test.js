
import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch');

describe('Test in <MultipleCustomHooks />', () => {
    
    test('should show correctly', () => {
        
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);

        expect( wrapper ).toMatchSnapshot();

    });

    test('should show the info', () => {

        useFetch.mockReturnValue({
            data: [{
                author: 'Nahuel',
                quote: "Hola mundo"
            }],

            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);

        expect(wrapper).toMatchSnapshot();
    });

});