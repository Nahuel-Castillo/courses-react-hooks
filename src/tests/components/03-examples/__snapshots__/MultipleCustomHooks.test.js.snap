// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <MultipleCustomHooks /> should show correctly 1`] = `
<div>
  <h2>
     BrekingBad Quotes 
  </h2>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</div>
`;

exports[`Test in <MultipleCustomHooks /> should show the info 1`] = `
<div>
  <h2>
     BrekingBad Quotes 
  </h2>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-3"
    >
       
      Hola mundo
       
    </p>
    <footer
      className="blockquote-footer"
    >
       
      Nahuel
        
    </footer>
  </blockquote>
  <hr />
  <button
    className="btn btn-primary m-2"
    disabled={true}
    onClick={[Function]}
  >
    Previous quote
  </button>
  <button
    className="btn btn-primary m-2"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
